Google Data Analytics Certificate Capstone Project

Introduction - This case study is my capstone project for the Google Data Analytics Certificate. It involves analysis of historical data for a fictional company, Cyclistic, a bike sharing company located in Chicago, to make recommendations for an upcoming marketing campaign. Although the company and scenario are fictitious, the data used for this project are real data collected between January 2021 – December 2021 from a bike share program in Chicago. In this project I am assuming the role of the junior analyst.

Scenario - The director of marketing believes the company’s future success depends on maximizing the number of annual memberships. Therefore, your team wants to understand how casual riders and annual members use Cyclistic bikes differently. From these insights, your team will design a new marketing strategy to convert casual riders into annual members. But first, Cyclistic executives must approve your recommendations, so they must be backed up with compelling data insights and professional data visualizations.

Characters and teams - 
● Cyclistic: A bike-share program that features more than 5,800 bicycles and 600 docking stations. Cyclistic sets itself apart by also offering reclining bikes, hand tricycles, and cargo bikes, making bike-share more inclusive to people with disabilities and riders who can’t use a standard two-wheeled bike. The majority of riders opt for traditional bikes; about 8% of riders use the assistive options. Cyclistic users are more likely to ride for leisure, but about 30% use them to commute to work each day.
● Lily Moreno: The director of marketing and your manager. Moreno is responsible for the development of campaigns
and initiatives to promote the bike-share program. These may include email, social media, and other channels.
● Cyclistic marketing analytics team: A team of data analysts who are responsible for collecting, analyzing, and
reporting data that helps guide Cyclistic marketing strategy. You joined this team six months ago and have been busy learning about Cyclistic’s mission and business goals — as well as how you, as a junior data analyst, can help Cyclistic
achieve them.
● Cyclistic executive team: The notoriously detail-oriented executive team will decide whether to approve the
recommended marketing program.

Business Objective - 
To increase profitability by converting casual riders to annual members via a targeted marketing campaign.
Task - 
How do annual members and casual riders use Cyclistic bikes differently?

In [4]:
#import the pandas library
import pandas as pd

In [5]:
#load the data for first month
df = pd.read_csv("D:\\Study\\Google Analytics\\cours 8\\cyclist\\202101-divvy-tripdata.csv", index_col=False)

In [6]:
#loading data for months 2 to 9
for i in range(2,10):
    path = "D:\\Study\\Google Analytics\\cours 8\\cyclist\\20210"+str(i)+"-divvy-tripdata.csv"
    dftemp = pd.read_csv(path, index_col=False)
    df = pd.concat([df,dftemp],axis = 0)

In [7]:
#loading data for months 11 and 12
for i in range(10,13):
    path = "D:\\Study\\Google Analytics\\cours 8\\cyclist\\2021"+str(i)+"-divvy-tripdata.csv"
    dftemp = pd.read_csv(path, index_col=False)
    df = pd.concat([df,dftemp],axis = 0)

In [8]:
df.head()

ride_id  rideable_type           started_at             ended_at  \
0  E19E6F1B8D4C42ED  electric_bike  2021-01-23 16:14:19  2021-01-23 16:24:44   
1  DC88F20C2C55F27F  electric_bike  2021-01-27 18:43:08  2021-01-27 18:47:12   
2  EC45C94683FE3F27  electric_bike  2021-01-21 22:35:54  2021-01-21 22:37:14   
3  4FA453A75AE377DB  electric_bike  2021-01-07 13:31:13  2021-01-07 13:42:55   
4  BE5E8EB4E7263A0B  electric_bike  2021-01-23 02:24:02  2021-01-23 02:24:45   

           start_station_name start_station_id end_station_name  \
0  California Ave & Cortez St            17660              NaN   
1  California Ave & Cortez St            17660              NaN   
2  California Ave & Cortez St            17660              NaN   
3  California Ave & Cortez St            17660              NaN   
4  California Ave & Cortez St            17660              NaN   

  end_station_id  start_lat  start_lng  end_lat  end_lng member_casual  
0            NaN  41.900341 -87.696743    41.89   -87.72        member  
1            NaN  41.900333 -87.696707    41.90   -87.69        member  
2            NaN  41.900313 -87.696643    41.90   -87.70        member  
3            NaN  41.900399 -87.696662    41.92   -87.69        member  
4            NaN  41.900326 -87.696697    41.90   -87.70        casual

In [9]:
len(df)

5595063

In [10]:
df.dtypes

ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object

The table has 13 columns and there are total 5595063 rows. The data type of columns started_at and ended_at is object which needs to be changed to datetime. No need to change datatype of rest of the columns.

In [11]:
#the data types for start and end time of ride is object time, need to convert it to datetime
df[['started_at','ended_at']] = df[['started_at','ended_at']].astype('datetime64')
df.dtypes

ride_id                       object
rideable_type                 object
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name            object
start_station_id              object
end_station_name              object
end_station_id                object
start_lat                    float64
start_lng                    float64
end_lat                      float64
end_lng                      float64
member_casual                 object
dtype: object

In [12]:
#clean string data, remove preceding or trailing extra spaces
df_obj = df.select_dtypes(['object'])
df[df_obj.columns] = df_obj.apply(lambda x: x.str.strip())
df.head()

ride_id  rideable_type          started_at            ended_at  \
0  E19E6F1B8D4C42ED  electric_bike 2021-01-23 16:14:19 2021-01-23 16:24:44   
1  DC88F20C2C55F27F  electric_bike 2021-01-27 18:43:08 2021-01-27 18:47:12   
2  EC45C94683FE3F27  electric_bike 2021-01-21 22:35:54 2021-01-21 22:37:14   
3  4FA453A75AE377DB  electric_bike 2021-01-07 13:31:13 2021-01-07 13:42:55   
4  BE5E8EB4E7263A0B  electric_bike 2021-01-23 02:24:02 2021-01-23 02:24:45   

           start_station_name start_station_id end_station_name  \
0  California Ave & Cortez St            17660              NaN   
1  California Ave & Cortez St            17660              NaN   
2  California Ave & Cortez St            17660              NaN   
3  California Ave & Cortez St            17660              NaN   
4  California Ave & Cortez St            17660              NaN   

  end_station_id  start_lat  start_lng  end_lat  end_lng member_casual  
0            NaN  41.900341 -87.696743    41.89   -87.72        member  
1            NaN  41.900333 -87.696707    41.90   -87.69        member  
2            NaN  41.900313 -87.696643    41.90   -87.70        member  
3            NaN  41.900399 -87.696662    41.92   -87.69        member  
4            NaN  41.900326 -87.696697    41.90   -87.70        casual

In [13]:
df.count()

ride_id               5595063
rideable_type         5595063
started_at            5595063
ended_at              5595063
start_station_name    4904254
start_station_id      4904257
end_station_name      4855893
end_station_id        4855893
start_lat             5595063
start_lng             5595063
end_lat               5590292
end_lng               5590292
member_casual         5595063
dtype: int64

The count function shows count of values in each col (count function ignores NaN). It shows there are some missing in a number of columns.

In [14]:
# to check different bike types
df['rideable_type'].unique()

array(['electric_bike', 'classic_bike', 'docked_bike'], dtype=object)

In [15]:
# unique values to check if there are only two member types
df['member_casual'].unique()

array(['member', 'casual'], dtype=object)

In [16]:
# check missing values in each column
df.isna().sum()

ride_id                    0
rideable_type              0
started_at                 0
ended_at                   0
start_station_name    690809
start_station_id      690806
end_station_name      739170
end_station_id        739170
start_lat                  0
start_lng                  0
end_lat                 4771
end_lng                 4771
member_casual              0
dtype: int64

There are a number of columns that have missing values. The rows containing missing values have to be removed. As ride_id uniquely identifies a ride, rows with duplicate ride_id also have to be removed.

In [17]:
#remove duplicates based on ride_id
df.drop_duplicates(subset="ride_id", keep=False, inplace=True)

In [18]:
#remove rows with missing values start_station, end_station, start/end lat/long
df2 = df.dropna()
df2.head()

ride_id  rideable_type          started_at            ended_at  \
9   B9F73448DFBE0D45   classic_bike 2021-01-24 19:15:38 2021-01-24 19:22:51   
10  457C7F4B5D3DA135  electric_bike 2021-01-23 12:57:38 2021-01-23 13:02:10   
11  57C750326F9FDABE  electric_bike 2021-01-09 15:28:04 2021-01-09 15:37:51   
12  4D518C65E338D070  electric_bike 2021-01-09 15:28:57 2021-01-09 15:37:54   
13  9D08A3AFF410474D   classic_bike 2021-01-24 15:56:59 2021-01-24 16:07:08   

            start_station_name start_station_id            end_station_name  \
9   California Ave & Cortez St            17660      Wood St & Augusta Blvd   
10  California Ave & Cortez St            17660  California Ave & North Ave   
11  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
12  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
13  California Ave & Cortez St            17660      Wood St & Augusta Blvd   

   end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  
9             657  41.900363 -87.696704  41.899181 -87.672200        member  
10          13258  41.900406 -87.696733  41.910435 -87.696890        member  
11            657  41.900374 -87.696688  41.899180 -87.672178        casual  
12            657  41.900379 -87.696716  41.899149 -87.672177        casual  
13            657  41.900363 -87.696704  41.899181 -87.672200        casual

In [19]:
#remove rides tha lasted less than 1 min or more than 1 day (as suggested in project document)
#first create ride_length column by subtracting start datetime from end datetime
df2['ride_length'] = df2['ended_at'] - df2['started_at']
df2.head()

C:\Users\ybhal\AppData\Local\Temp\ipykernel_6548\4151079827.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['ride_length'] = df2['ended_at'] - df2['started_at']


ride_id  rideable_type          started_at            ended_at  \
9   B9F73448DFBE0D45   classic_bike 2021-01-24 19:15:38 2021-01-24 19:22:51   
10  457C7F4B5D3DA135  electric_bike 2021-01-23 12:57:38 2021-01-23 13:02:10   
11  57C750326F9FDABE  electric_bike 2021-01-09 15:28:04 2021-01-09 15:37:51   
12  4D518C65E338D070  electric_bike 2021-01-09 15:28:57 2021-01-09 15:37:54   
13  9D08A3AFF410474D   classic_bike 2021-01-24 15:56:59 2021-01-24 16:07:08   

            start_station_name start_station_id            end_station_name  \
9   California Ave & Cortez St            17660      Wood St & Augusta Blvd   
10  California Ave & Cortez St            17660  California Ave & North Ave   
11  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
12  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
13  California Ave & Cortez St            17660      Wood St & Augusta Blvd   

   end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  \
9             657  41.900363 -87.696704  41.899181 -87.672200        member   
10          13258  41.900406 -87.696733  41.910435 -87.696890        member   
11            657  41.900374 -87.696688  41.899180 -87.672178        casual   
12            657  41.900379 -87.696716  41.899149 -87.672177        casual   
13            657  41.900363 -87.696704  41.899181 -87.672200        casual   

       ride_length  
9  0 days 00:07:13  
10 0 days 00:04:32  
11 0 days 00:09:47  
12 0 days 00:08:57  
13 0 days 00:10:09

In [20]:
#convert ride length to seconds
def timedelta_to_seconds(timedelta):
    return int(timedelta.total_seconds())
df2['ride_seconds'] = df2['ride_length'].apply(timedelta_to_seconds)

C:\Users\ybhal\AppData\Local\Temp\ipykernel_6548\3665462807.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['ride_seconds'] = df2['ride_length'].apply(timedelta_to_seconds)


In [21]:
df2.head()

ride_id  rideable_type          started_at            ended_at  \
9   B9F73448DFBE0D45   classic_bike 2021-01-24 19:15:38 2021-01-24 19:22:51   
10  457C7F4B5D3DA135  electric_bike 2021-01-23 12:57:38 2021-01-23 13:02:10   
11  57C750326F9FDABE  electric_bike 2021-01-09 15:28:04 2021-01-09 15:37:51   
12  4D518C65E338D070  electric_bike 2021-01-09 15:28:57 2021-01-09 15:37:54   
13  9D08A3AFF410474D   classic_bike 2021-01-24 15:56:59 2021-01-24 16:07:08   

            start_station_name start_station_id            end_station_name  \
9   California Ave & Cortez St            17660      Wood St & Augusta Blvd   
10  California Ave & Cortez St            17660  California Ave & North Ave   
11  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
12  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
13  California Ave & Cortez St            17660      Wood St & Augusta Blvd   

   end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  \
9             657  41.900363 -87.696704  41.899181 -87.672200        member   
10          13258  41.900406 -87.696733  41.910435 -87.696890        member   
11            657  41.900374 -87.696688  41.899180 -87.672178        casual   
12            657  41.900379 -87.696716  41.899149 -87.672177        casual   
13            657  41.900363 -87.696704  41.899181 -87.672200        casual   

       ride_length  ride_seconds  
9  0 days 00:07:13           433  
10 0 days 00:04:32           272  
11 0 days 00:09:47           587  
12 0 days 00:08:57           537  
13 0 days 00:10:09           609

In [22]:
#remove rows where the ride length is less than 60 seconds or more than 86400 (24 hours) seconds
df3 = df2[(df2['ride_seconds']<86400) | (df2['ride_seconds']>60)]
df3.reset_index()
df3.head()

ride_id  rideable_type          started_at            ended_at  \
9   B9F73448DFBE0D45   classic_bike 2021-01-24 19:15:38 2021-01-24 19:22:51   
10  457C7F4B5D3DA135  electric_bike 2021-01-23 12:57:38 2021-01-23 13:02:10   
11  57C750326F9FDABE  electric_bike 2021-01-09 15:28:04 2021-01-09 15:37:51   
12  4D518C65E338D070  electric_bike 2021-01-09 15:28:57 2021-01-09 15:37:54   
13  9D08A3AFF410474D   classic_bike 2021-01-24 15:56:59 2021-01-24 16:07:08   

            start_station_name start_station_id            end_station_name  \
9   California Ave & Cortez St            17660      Wood St & Augusta Blvd   
10  California Ave & Cortez St            17660  California Ave & North Ave   
11  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
12  California Ave & Cortez St            17660      Wood St & Augusta Blvd   
13  California Ave & Cortez St            17660      Wood St & Augusta Blvd   

   end_station_id  start_lat  start_lng    end_lat    end_lng member_casual  \
9             657  41.900363 -87.696704  41.899181 -87.672200        member   
10          13258  41.900406 -87.696733  41.910435 -87.696890        member   
11            657  41.900374 -87.696688  41.899180 -87.672178        casual   
12            657  41.900379 -87.696716  41.899149 -87.672177        casual   
13            657  41.900363 -87.696704  41.899181 -87.672200        casual   

       ride_length  ride_seconds  
9  0 days 00:07:13           433  
10 0 days 00:04:32           272  
11 0 days 00:09:47           587  
12 0 days 00:08:57           537  
13 0 days 00:10:09           609

In [24]:
#export data to csv for visualization in powerbi
df3.to_csv("cleaned_tripdata.csv",index=False)